src/Microsoft.IdentityModel.Tokens/ECDsaSecurityKey.cs

```diff
@@ -42,30 +42,20 @@ public ECDsaSecurityKey(byte[] blob, CngKeyBlobFormat blobFormat)
 
             CngKey = CngKey.Import(blob, blobFormat);
             BlobFormat = blobFormat;
+            HasPrivateKey = BlobFormat.Format == CngKeyBlobFormat.EccPrivateBlob.Format || BlobFormat.Format == CngKeyBlobFormat.GenericPrivateBlob.Format;
+            HasPublicKey = HasPrivateKey || BlobFormat.Format == CngKeyBlobFormat.EccPublicBlob.Format || BlobFormat.Format == CngKeyBlobFormat.GenericPublicBlob.Format;
         }
 
-        public override bool HasPrivateKey
+        public ECDsaSecurityKey(ECDsa ecdsa)
         {
-            get
-            {
-                return (BlobFormat.Format == CngKeyBlobFormat.EccPrivateBlob.Format || BlobFormat.Format == CngKeyBlobFormat.GenericPrivateBlob.Format);
-            }
-        }
+            if (ecdsa == null)
+                throw LogHelper.LogArgumentNullException("ecdsa");
 
-        public override bool HasPublicKey
-        {
-            get
-            {
-                return (HasPrivateKey || BlobFormat.Format == CngKeyBlobFormat.EccPublicBlob.Format || BlobFormat.Format == CngKeyBlobFormat.GenericPublicBlob.Format);
-            }
-        }
+            ECDsa = ecdsa;
 
-        public override SignatureProvider GetSignatureProvider(string algorithm, bool verifyOnly)
-        {
-            if (verifyOnly)
-                return SignatureProviderFactory.CreateForVerifying(this, algorithm);
-            else
-                return SignatureProviderFactory.CreateForSigning(this, algorithm);
+            // fake signing to determine if the ecdsa instance has the private key or not is a costly operation especially in case of HSM. We return true by default in that case, it will fail later at the time of signing or decrypting.
+            HasPrivateKey = true;
```

Comment:
Another alternative is to use the approach we had in System.IdentityModel.Tokens to use an    `enum PrivateKeyStatus
        {
            AvailabilityNotDetermined,
            HasPrivateKey,
            DoesNotHavePrivateKey
        }`
I agree that its preferable to let the user know the exact status rather than setting an incorrect value. @brentschmaltz